modules/nw-ovn-k-anp-best-practices.adoc

```diff
@@ -0,0 +1,36 @@
+/ Module included in the following assemblies:
+//
+// * list of assemblies where this module is included:
+// networking/network_security/AdminNetworkPolicy/nw-ovn-k-anp-recommended-practices.adoc
+
+:_mod-docs-content-type: REFERENCE
+[id="anp-best-practices_{context}"]
+= Designing AdminNetworkPolicy
+
+When building `AdminNetworkPolicy` (ANP), you should consider the following when creating your policies:
+
+//This will need to be added in at 4.16.1
+
+//* You should create `AdminNetworkPolicy` at different priorities so that precedence is well defined. It is not guaranteed which policy will take precedence when overlapping ANP are created.
+
+* There is only one `BaselineAdminNetworkPolicy` for the cluster so be careful in defining it and you might consider using it as a default guardrail policy.
+
+* Since `0-100` is the supported priority range, it is always better when designing your ANP to use a middle range like `30-70` because this leaves some placeholder
+for priorities before and after. Even in the middle range, you might want to leave gaps so that as your infrastructure requirements evolve over time, you are able to insert new ANPs when needed at the right priority level. If you pack your ANPs, then you might need to recreate all of them to accommodate any changes in the future.
+
+* `networks` peer in egress rule is meant to select external destinations. It is recommended to use `namespaces` and `pods` peers to select in-cluster peers. Do not use `networks` for podIP level matching.
+
+* When using `0.0.0.0/0` to create a strong `DENY` policy, it is recommended that you check to make sure that you have higher priority `Allow` rules for essential traffic like towards DNS and KAPI-server. This ensures that you do not lock up your entire cluster leaving it in a non-functional state.
+
+* `Pass` is a special action used to delegate policy decision to the `NetworkPolicy` layer. If you might consider using this action when considering the interoperability between ANP and your `NetworkPolicy`.
+
+* Each rule can have up to 100 peers. Every peer is added to the same `Address_Set` object. Selectors across multiple rules should not overlap. This ensures that that the same IPs do not appear in multiple `Address_Set`, which can cause performance and scale scale limitations due to an OVN-Kubernetes limitation.
+
+* Using `namedPorts` is not performance efficient since each name has to get evaluated against the `subject` of the policy and `peer` pod's container ports. You might use this with caution in larger clusters where you have many pods and each pod has many containers.
```

Comment:
Again, this is mixing in too much low-level information. People don't need to know the low-level details of why it is inefficient. Just tell them what sorts of policies it is OK to use namedPorts in, and what kinds of policies it is not OK, and just _how_ not OK it is. (Is it 10 times less efficient? 10,000 times?)

"has many containers" does not seem relevant at all.

Reply:
@tssurya this one I definitely need help with. cc @asood-rh

Reply:
Using ports (namedport) creates 3 ACLs, one per per protocol, 3ACLs one per protocol if ports (PortNumber, PortRange) are specified in the ANP. If both are used, results in 6 ACLs. So use wisely.

Reply:
@danwinship I'm not sure I know how we could even figure out how less efficient it is. I suppose the point here is more along the lines of what @asood-rh is saying which is it isn't a good idea to use `namedPorts` in conjunction with `PortNumber` and `PortRange`. I'm going to update this along those lines for another round of feedback. It may be that we should just take this out.